# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [2]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143
The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! yo

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [3]:
import problem_unittests as tests

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    unique_words = set(text)
    int_to_vocab = dict(enumerate(unique_words))
    vocab_to_int = {v: k for k, v in int_to_vocab.items()}
    
    # return tuple
    return (vocab_to_int, int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

Tests Passed


### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [4]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
        
    return {
        '.': '||period||',
        ',': '||comma||',
        '"': '||double_quote||',
        ';': '||semicolon||',
        '!': '||exclamation||',
        '?': '||question||',
        '(': '||open_paren||',
        ')': '||close_paren||',
        '-': '||dash||',
        '\n': '||newline||',
    }

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to look at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

In [8]:
from torch.utils.data import TensorDataset, DataLoader


def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
    x = []
    y = []
    for i in range(len(words) - sequence_length - 1):
        x.append(words[i:i+sequence_length])
        y.append(words[i+sequence_length])
    
    ds = TensorDataset(torch.tensor(x),  torch.tensor(y))
    dl = DataLoader(ds, batch_size=batch_size)
    
    # return a dataloader
    return dl

# there is no test for this function, but you are encouraged to create
# print statements and tests of your own


### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [9]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 0,  1,  2,  3,  4],
        [ 1,  2,  3,  4,  5],
        [ 2,  3,  4,  5,  6],
        [ 3,  4,  5,  6,  7],
        [ 4,  5,  6,  7,  8],
        [ 5,  6,  7,  8,  9],
        [ 6,  7,  8,  9, 10],
        [ 7,  8,  9, 10, 11],
        [ 8,  9, 10, 11, 12],
        [ 9, 10, 11, 12, 13]])

torch.Size([10])
tensor([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

In [10]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        # TODO: Implement function
        
        # set class variables
        self.vocab_size = vocab_size
        self.output_size = output_size
        self.embedding_dim = embedding_dim
        self.hidden_dim = hidden_dim
        self.n_layers = n_layers
        
        # define model layers
        self.embedding = nn.Embedding(self.vocab_size, self.embedding_dim)
        self.lstm = nn.LSTM(self.embedding_dim, self.hidden_dim, n_layers, batch_first=True, dropout=dropout)
        self.dropout = nn.Dropout(dropout)
        self.fc = nn.Linear(in_features=self.hidden_dim, out_features=self.vocab_size)
        self.softmax = nn.Softmax(dim=1)
    
    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        # TODO: Implement function
        batch_size, seq_len = nn_input.shape
        embed = self.embedding(nn_input)
        lstm_out, hidden = self.lstm(embed, hidden)
#         out = self.dropout(out)
        lstm_out_stacked = lstm_out.contiguous().view(-1, self.hidden_dim)
        fc_out = self.fc(lstm_out_stacked)
        sm_out = fc_out#self.softmax(fc_out)
        sm_out_resized = sm_out.view(batch_size, -1, self.output_size)
        out = sm_out_resized[:, -1]
#         print('nn_input {}, embed {}, lstm_out {}, fc_out {}, sm_out {}, out {}'.format(
#             nn_input.shape, embed.shape, lstm_out.shape, fc_out.shape, sm_out.shape, out.shape))

        # return one batch of output word scores and the hidden state
        return out, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # Implement function
        
        # initialize hidden state with zero weights, and move to GPU if available
        self.batch_size = batch_size
        dim = [self.n_layers, self.batch_size, self.hidden_dim]
        h = torch.zeros(dim)
        c = torch.zeros(dim)
        if train_on_gpu:
            h = h.cuda()
            c = c.cuda()
        
    
        return h, c

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

In [11]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param rnn: The PyTorch Module that holds the neural network
    :param optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """
    
    # TODO: Implement Function
    
    # move data to GPU, if available
    if train_on_gpu:
        h_0, c_0 = hidden
        inp, target, hidden = inp.cuda(), target.cuda(), (h_0.cuda(), c_0.cuda())
    
    # perform backpropagation and optimization
    out, hidden = rnn(inp, hidden)
    loss = criterion(out, target)
    hidden = tuple([each.data for each in hidden])
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    # return the loss over a batch and the hidden state produced by our model
    return loss.item(), hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [12]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of unique tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

### Grid Search for hyperparameter optimization

The following cell performs a grid search on a small subset of data, 
to find the best candidates for the following values:
* Embedding dimension
* Hidden layer dimension
* Sequence length
* Number of hidden layers
* Batch size

In [13]:
from random import random
from time import time, localtime, strftime
import pickle

In [22]:
learning_rate = 0.001
vocab_size = len(vocab_to_int)
output_size = vocab_size

# basically a shameless copy-paste of the provided `train_rnn` function, with modified outputs
def __train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, train_loader):
    batch_losses = []
    
    rnn.train()
    t0 = time()
    for epoch_i in range(1, n_epochs + 1):
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)
        
        t_diff = time() - t0
#         print('Epoch {}/{} took: {:.2f} s'.format(epoch_i, n_epochs, t_diff))

    # drop initial losses
    batch_losses = batch_losses[len(batch_losses)//2:]
    return rnn, np.average(batch_losses)

def __subbatch(int_text, sample_size_perc, sequence_length, train_batch_size, valid_batch_size=8):
    words_total = len(int_text)
    sample_size = int(words_total * sample_size_perc / 100)
    valid_size = sample_size // 5
    max_idx = words_total - sample_size - valid_size
    start_idx = int(random() * max_idx)
    start_train = start_idx
    end_train = start_train + sample_size
    start_valid = end_train
    end_valid = start_valid + valid_size
    train_sample = int_text[start_train: end_train]
    valid_sample = int_text[start_valid: end_valid]
#     print(f'total_text: {len(int_text)}, train_sample: {len(train_sample)}, valid_sample: {len(valid_sample)}')
#     print(f'start_idx={start_idx}, sample size {len(sample)}')

    train_loader = batch_data(train_sample, sequence_length, train_batch_size)
    valid_loader = batch_data(valid_sample, sequence_length, valid_batch_size)
    
    return train_loader, valid_loader

def __train(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, train_loader, n_epochs=5):
# #     print(f'Train, n_layers: {n_layers}, emb_dim: {embedding_dim}, hid_dim: {hidden_dim}, {len(train_loader)} samples')
    rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5)
    if train_on_gpu:
        rnn.cuda()

    # defining loss and optimization functions for training
    optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
    criterion = nn.CrossEntropyLoss()

    # training the model
    return __train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, train_loader)

def __valid(rnn, valid_loader, criterion, batch_size):
    rnn.eval()
    n_batches = len(valid_loader.dataset)//batch_size
#     print(f'valid: {n_batches} batches of size {batch_size}')
    losses = []
    
    for i, (x, y) in enumerate(valid_loader):
        if i >= n_batches:
            break
            
        with torch.no_grad():  
            hidden = rnn.init_hidden(batch_size)
            if train_on_gpu:
                h_0, c_0 = hidden
#                 print(f'Hidden: {h_0.shape}')
                x, y, hidden = x.cuda(), y.cuda(), (h_0.cuda(), c_0.cuda())
            
            res, _ = rnn(x, hidden)
            loss = criterion(res, y)
            losses.append(loss.item())
    
    return np.average(losses)
        

def __make_grid():
    hidden_dims = [200, 400, 800]
    layers = [2,4]
    embedding_dims = [100,200,400]
    sequence_lengths = [20, 40, 80]
    batch_sizes = [64, 128]
    perms = []
    for nl in layers:
        for hd in hidden_dims:
            for ed in embedding_dims:
                for sl in sequence_lengths:
                    for bs in batch_sizes:
                        perms.append((nl, hd, ed, sl, bs))
    return perms



In [ ]:
results = []

sample_perc = 5
start_time = strftime('%Y%m%d_%H%M%S', localtime())
file_prefix = 'grid_results_{}_{}'.format(sample_perc, start_time)
grid_text_file = f'{file_prefix}.txt'
grid_results_file = f'{file_prefix}.p'
print(f'Files: {grid_text_file}, {grid_results_file}')

perms = __make_grid()
print(f'Eval for {len(perms)} hyperparams value sets')
for perm_i, perm in enumerate(perms):
    n_layers, h_dim, emb_dim, seq_len, batch_size = perm
#     print(n_layers, h_dim, emb_dim, seq_len, batch_size)
    train_loader, valid_loader = __subbatch(int_text, sample_perc, seq_len, batch_size)
#     print('Start permutation {}/{}: {}, train sample: {}, valid sample: {}' \
#           .format(perm_i, len(perms), perm, len(train_loader.dataset), len(valid_loader.dataset)))
    t0 = time()
    trained, train_loss = __train(vocab_size, output_size, emb_dim, h_dim, n_layers, train_loader)
    t1 = time()
#     print('Finished training {} in {:.2f} s, train loss {}'.format(perm, t1 - t0, train_loss))
    valid_loss = __valid(trained, valid_loader, nn.CrossEntropyLoss(), 8)
    t2 = time()
#     print('Finished validating {} in {:.2f}, valid loss {}'.format(perm, t2 - t1, valid_loss))
    results.append((n_layers, h_dim, emb_dim, seq_len, batch_size, train_loss, valid_loss))
    
    with open(grid_text_file, 'a') as res_file:
        output = 'time={:.2f} s, n_layers={} h_dim={} emb_dim={} seq_len={} batch_size={}, train_loss={:.3f}, valid_loss={:.3f}\n' \
            .format(t2-t0, n_layers, h_dim, emb_dim, seq_len, batch_size, train_loss, valid_loss)
        res_file.write(output)
#         print(output)

pickle.dump(results, open(grid_results_file, 'wb'))


Files: grid_results_5_20211031_150705.txt, grid_results_5_20211031_150705.p
Eval for 108 hyperparams value sets


After testing 108 hyperparameter value sets on 5% of data,
5 best (by validation error) are picked and tested on 20% of data

In [25]:
with open(grid_results_file, 'rb') as in_file:
    _results = pickle.load(in_file)

best_5_hyp_params = sorted(_results, key=lambda t: t[6])[:5]

In [33]:
import os.path
import os

sample_perc = 20
start_time = strftime('%Y%m%d_%H%M%S', localtime())
file_prefix = 'grid_results_{}_{}'.format(sample_perc, start_time)
grid_text_file = f'{file_prefix}.txt'
grid_results_file = f'{file_prefix}.p'
print(f'Files: {grid_text_file}, {grid_results_file}')

Files: grid_results_20_20211101_093618.txt, grid_results_20_20211101_093618.p


In [35]:
results = []
for n_layers, h_dim, emb_dim, seq_len, batch_size, _, _ in best_5_hyp_params:
    t0 = time()
    train_loader, valid_loader = __subbatch(int_text, sample_perc, seq_len, batch_size)
    trained, train_loss = __train(vocab_size, output_size, emb_dim, h_dim, n_layers, train_loader, n_epochs=10)
    valid_loss = __valid(trained, valid_loader, nn.CrossEntropyLoss(), 8)
    t1 = time()
    results.append((n_layers, h_dim, emb_dim, seq_len, batch_size, train_loss, valid_loss))
    
    with open(best_5_file_txt, 'a') as res_file:
        output = 'time={:.2f} s, n_layers={} h_dim={} emb_dim={} seq_len={} batch_size={}, train_loss={:.3f}, valid_loss={:.3f}\n' \
            .format(t1-t0, n_layers, h_dim, emb_dim, seq_len, batch_size, train_loss, valid_loss)
        res_file.write(output)
        print(output)

pickle.dump(results, open(best_5_file_res, 'wb'))


time=2072.87 s, n_layers=2 h_dim=400 emb_dim=400 seq_len=80 batch_size=128, train_loss=2.900, valid_loss=5.150

time=510.76 s, n_layers=2 h_dim=200 emb_dim=100 seq_len=40 batch_size=64, train_loss=3.172, valid_loss=4.973

time=535.16 s, n_layers=2 h_dim=200 emb_dim=200 seq_len=40 batch_size=64, train_loss=3.172, valid_loss=5.182

time=688.55 s, n_layers=2 h_dim=400 emb_dim=200 seq_len=40 batch_size=128, train_loss=2.975, valid_loss=5.412

time=1412.05 s, n_layers=2 h_dim=400 emb_dim=200 seq_len=80 batch_size=64, train_loss=2.784, valid_loss=5.835



In [43]:
with open(grid_results_file, 'rb') as in_file:
    _results = pickle.load(in_file)

best_2_hyp_params = sorted(_results, key=lambda t: t[6])[0]
best_2_hyp_params

(2, 200, 100, 40, 64, 3.172282147373313, 4.973261087898721)

In [ ]:
import os.path
import os

sample_perc = 20
start_time = strftime('%Y%m%d_%H%M%S', localtime())
file_prefix = 'grid_results_{}_{}'.format(sample_perc, start_time)
grid_text_file = f'{file_prefix}.txt'
grid_results_file = f'{file_prefix}.p'
print(f'Files: {grid_text_file}, {grid_results_file}')

The following hyperparameters were set to:
* hidden layers: 2
* embedding dim: 100
* hidden dim: 200
* sequence length: 40
* batch size: 64

Higher values for hidden dimension, embedding dimension and sequence length resulted in quicker convergence of training error, but caused overtraining. Values above gave the lowest validation error with acceptable training error (< 3.5)

In [37]:
# Data params
# Sequence Length
sequence_length = 40  # of words in a sequence
# Batch Size
batch_size = 64

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [38]:
# Training parameters
# Number of Epochs
num_epochs = 20
# Learning Rate
learning_rate = .001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 100
# Hidden Dimension
hidden_dim = 200
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 500

In [39]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.3)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 20 epoch(s)...
Epoch:    1/20    Loss: 5.742737386703491

Epoch:    1/20    Loss: 5.138306752681732

Epoch:    1/20    Loss: 4.990906141281128

Epoch:    1/20    Loss: 4.814815385341644

Epoch:    1/20    Loss: 4.74476785326004

Epoch:    1/20    Loss: 4.603368194580078

Epoch:    1/20    Loss: 4.517326819419861

Epoch:    1/20    Loss: 4.568792717933655

Epoch:    1/20    Loss: 4.630016386985779

Epoch:    1/20    Loss: 4.463040771007538

Epoch:    1/20    Loss: 4.558808767318726

Epoch:    1/20    Loss: 4.600115958690643

Epoch:    1/20    Loss: 4.51828777885437

Epoch:    1/20    Loss: 4.453363593578339

Epoch:    1/20    Loss: 4.467135589599609

Epoch:    1/20    Loss: 4.271032493591308

Epoch:    1/20    Loss: 4.301664689064026

Epoch:    1/20    Loss: 4.382629855632782

Epoch:    1/20    Loss: 4.317809194087983

Epoch:    1/20    Loss: 4.241667837142944

Epoch:    1/20    Loss: 4.365644994735717

Epoch:    1/20    Loss: 4.440709570407868

Epoch:    1/20    Loss: 4.45

### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** 

1. Using grid search to check multiple values of
   `(n_layers, hidden_dim, embedding_dim, sequence_length, batch_size)` on small subset of data (5%), then picking    the best permutation and using it to train on full data.
2. Training the best 5 models from `P.1` on 20% of data
3. Training the model using the best hyperparameters from `P.2` using full data and increased number of epochs.

Models were chosen based on validation error. In general, increasing values of hyperparameters
ultimately led to overtraining. Using shorter sequences and lower hidden/embedding dimensions
led to better results, measured by validation error.

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [44]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [46]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        if(train_on_gpu):
            current_seq = current_seq.cpu() # move to cpu
        # the generated word becomes the next "current sequence" and the cycle can continue
        if train_on_gpu:
            current_seq = current_seq.cpu()
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [47]:
# run the cell multiple times to get different results!
gen_length = 400 # modify the length to your preference
prime_word = 'jerry' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

jerry:

jerry: yeah, it's a pet- 7 lot of the air.

george: oh, that's right.

george: so what do you do?

george: what is the difference? you know, he wants to see a lot of other.

jerry: oh, that's the right. i can't. it's rather delicious, but you know, it's a very nice day, but he wants to go to the movies.

george: so you're not the last person.

george: well, it's a very good friend.

jerry: i know.

elaine: oh, i forgot.

jerry: hey! hey.

george: you know, you can't do that.

elaine: what? you don't want to see me?

george: no, i can't. i'm going to the stand. i was just going to see what they have.

jerry: so you're going to go?

george: oh, yeah.

jerry: oh, that's right.

elaine: oh, yeah, yeah, yeah yeah.

george: yeah, well, i'm sorry to be a very nice guy.

jerry: oh, it's not the point.

elaine: yeah.

jerry: so, how was your parents that?

elaine: yeah, yeah.

jerry: oh, hi, mr. seinfeld.

kramer:(handing jerry to the door) hey, hey. hey, hey.

elaine: oh, hi buddy.

je

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [48]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.